From f2658c137c308d3f392741486c1a7b7ce55eb6a4 Mon Sep 17 00:00:00 2001
From: Palanix <palanixyt@gmail.com>
Date: Fri, 27 May 2022 13:48:04 +0200
Subject: [PATCH] Fix swallow patch

---
 config.def.h |  6 ++--
 dwl.c        | 87 ++++++++++++++++++++++++++++++++++++++++++++++++++--
 2 files changed, 87 insertions(+), 6 deletions(-)

diff --git a/config.def.h b/config.def.h
index 4f131dd..8db27a2 100644
--- a/config.def.h
+++ b/config.def.h
@@ -10,11 +10,11 @@ static const float focuscolor[]     = {1.0, 0.0, 0.0, 1.0};
 static const char *tags[] = { "1", "2", "3", "4", "5", "6", "7", "8", "9" };
 
 static const Rule rules[] = {
-	/* app_id     title       tags mask     isfloating   monitor */
+	/* app_id     title       tags mask     isfloating  isterm          noswallow        monitor */
 	/* examples:
-	{ "Gimp",     NULL,       0,            1,           -1 },
+	{ "Gimp",     NULL,       0,            1,			0,				1,				-1 },
 	*/
-	{ "firefox",  NULL,       1 << 8,       0,           -1 },
+	{ "firefox",  NULL,       1 << 8,       0,          0,              1,              -1 },
 };
 
 /* layout(s) */
diff --git a/dwl.c b/dwl.c
index d0f5afc..1e897e1 100644
--- a/dwl.c
+++ b/dwl.c
@@ -89,7 +89,8 @@ typedef struct {
 } Button;
 
 typedef struct Monitor Monitor;
-typedef struct {
+typedef struct Client Client;
+struct Client{
 	/* Must be first */
 	unsigned int type; /* XDGShell or X11* */
 	struct wlr_scene_node *scene;
@@ -116,10 +117,12 @@ typedef struct {
 #endif
 	int bw;
 	unsigned int tags;
-	int isfloating, isurgent;
+	int isfloating, isurgent, isterm, noswallow;
 	uint32_t resize; /* configure serial of a pending resize */
+	pid_t pid;
+	Client *swallowing, *swallowedby;
 	int isfullscreen;
-} Client;
+};
 
 typedef struct {
 	uint32_t singular_anchor;
@@ -198,6 +201,8 @@ typedef struct {
 	const char *title;
 	unsigned int tags;
 	int isfloating;
+	int isterm;
+	int noswallow;
 	int monitor;
 } Rule;
 
@@ -294,6 +299,9 @@ static Monitor *xytomon(double x, double y);
 static struct wlr_scene_node *xytonode(double x, double y, struct wlr_surface **psurface,
 		Client **pc, LayerSurface **pl, double *nx, double *ny);
 static void zoom(const Arg *arg);
+static pid_t getparentprocess(pid_t p);
+static int isdescprocess(pid_t p, pid_t c);
+static Client *termforwin(Client *w);
 
 /* variables */
 static const char broken[] = "broken";
@@ -468,6 +476,8 @@ applyrules(Client *c)
 		if ((!r->title || strstr(title, r->title))
 				&& (!r->id || strstr(appid, r->id))) {
 			c->isfloating = r->isfloating;
+			c->isterm     = r->isterm;
+			c->noswallow  = r->noswallow;
 			newtags |= r->tags;
 			i = 0;
 			wl_list_for_each(m, &mons, link)
@@ -964,6 +974,8 @@ createnotify(struct wl_listener *listener, void *data)
 	c->surface.xdg = xdg_surface;
 	c->bw = borderpx;
 
+	wl_client_get_credentials(c->surface.xdg->client->client, &c->pid, NULL, NULL);
+
 	LISTEN(&xdg_surface->surface->events.commit, &c->commit, commitnotify);
 	LISTEN(&xdg_surface->events.map, &c->map, mapnotify);
 	LISTEN(&xdg_surface->events.unmap, &c->unmap, unmapnotify);
@@ -1240,6 +1252,48 @@ fullscreennotify(struct wl_listener *listener, void *data)
 	setfullscreen(c, fullscreen);
 }
 
+pid_t
+getparentprocess(pid_t p)
+{
+        unsigned int v = 0;
+
+        FILE *f;
+        char buf[256];
+        snprintf(buf, sizeof(buf) - 1, "/proc/%u/stat", (unsigned)p);
+
+        if (!(f = fopen(buf, "r")))
+                return 0;
+
+        fscanf(f, "%*u %*s %*c %u", &v);
+        fclose(f);
+
+       return (pid_t)v;
+}
+
+int
+isdescprocess(pid_t p, pid_t c)
+{
+        while (p != c && c != 0)
+                c = getparentprocess(c);
+
+        return (int)c;
+}
+
+Client *
+termforwin(Client *w)
+{
+        Client *c;
+
+        if (!w->pid || w->isterm || w->noswallow)
+                return NULL;
+
+        wl_list_for_each(c, &clients, link)
+                        if (c->isterm && !c->swallowing && c->pid && isdescprocess(c->pid, w->pid))
+                                return c;
+
+        return NULL;
+}
+
 void
 incnmaster(const Arg *arg)
 {
@@ -1416,6 +1470,21 @@ mapnotify(struct wl_listener *listener, void *data)
 		setfullscreen(c, 1);
 
 	c->mon->un_map = 1;
+	if (!c->noswallow) {
+			Client *p = termforwin(c);
+			if (p) {
+					c->swallowedby = p;
+					p->swallowing  = c;
+					wlr_scene_node_set_enabled(p->scene, 0);
+					wl_list_remove(&c->flink);
+					wl_list_remove(&c->link);
+					wl_list_insert(&p->flink, &c->flink);
+					wl_list_insert(&p->link, &c->link);
+					wl_list_remove(&p->flink);
+					wl_list_remove(&p->link);
+			}
+			arrange(c->mon);
+	}
 }
 
 void
@@ -2271,6 +2340,18 @@ unmapnotify(struct wl_listener *listener, void *data)
 		cursor_mode = CurNormal;
 		grabc = NULL;
 	}
+	if (c->swallowing) {
+			c->swallowing->swallowedby = NULL;
+			c->swallowing = NULL;
+	}
+
+	if (c->swallowedby) {
+			wl_list_insert(&c->link, &c->swallowedby->link);
+			wl_list_insert(&c->flink, &c->swallowedby->flink);
+			wlr_scene_node_set_enabled(c->swallowedby->scene, 1);
+			c->swallowedby->swallowing = NULL;
+			c->swallowedby = NULL;
+	}
 
 	if (c->mon)
 		c->mon->un_map = 1;
-- 
2.36.1

